FSW/controllers/controller.py

```diff
@@ -0,0 +1,168 @@
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            # reference angular velocity in pointing axis 
+            # ref_ctrl_states[:4]  = q_ref
+            # ref_ctrl_states[4:7] = tgt_ang_vel
+            ref_torque = np.zeros((3,))
+        elif pointing_mode == "3D-stabilized":
+            q_ref = rotmat2quat(Rot_mat)
+            ref_ctrl_states[:4] = q_ref
+            ref_torque = np.zeros((3,))
+        else:
+            raise ValueError(f"Unrecognized pointing mode: {pointing_mode}")
+            
+        return ref_torque, ref_ctrl_states
+    
+    def get_torque(self, date, ref_torque, ref_ctrl_states, est_ctrl_states, Idx):
+        # from the reference and estimated quaternions, get the error quaternion for control
+        if self.controller_algo == "Bcross":
+            mtb_torque = self.Bcrossctr.get_dipole_moment_command(est_ctrl_states, Idx)
+            return mtb_torque, []
+        elif self.controller_algo == "Lyapunov":
+            mtb_torque = self.lyapsunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseSP":
+            mtb_torque = self.basesunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseNP":
+            mtb_torque, rw_torque = self.nadirpointctr.get_dipole_moment_command_and_rw_torque(est_ctrl_states, Idx) 
+            return mtb_torque, rw_torque
+        else: 
+            raise ValueError(f"Unrecognized controller algorithm: {self.controller_algo}")
+        
+    
+    def allocate_torque(self, state, mtb_torque_cmd, rw_torque_cmd, Idx):
+        # Placeholder for actuator management
+        # if torque = B @ actuator_cmd
+        # actuator_cmd = Binv @ torque_cmd
+        
+     
+        self.allocation_mat = np.zeros((Idx["NU"],3))
+        self.allocation_mat[Idx["U"]["MTB_TORQUE"],:] = np.linalg.pinv(self.G_mtb_b.T)
+                
+        # np.linalg.pinv(B_mat)
+        actuator_cmd = np.zeros((Idx["NU"],1))
+        actuator_cmd[Idx["U"]["MTB_TORQUE"]] = np.linalg.pinv(self.G_mtb_b.T) @ mtb_torque_cmd
+        if rw_torque_cmd:
+            actuator_cmd[Idx["U"]["RW_TORQUE"]] = rw_torque_cmd # only one, bypass allocation matrix
+        # np.linalg.pinv(self.G_rw_b.T) @ rw_torque_cmd
+        # Normalize columns of the allocation matrix
+        # col_norms = np.linalg.norm(allocation_mat, axis=0)
+        # allocation_mat = allocation_mat / col_norms
+        
+        return actuator_cmd
+    
+    def run(self, date, est_world_states, Idx):
+        
+        self.est_world_states = est_world_states
+        # define slew profile
+        ref_torque, ref_ctrl_states = self.define_att_profile(date, self.pointing_mode, self.pointing_target, 
+                                                              self.tgt_ang_vel, est_world_states)
```

Comment:
do we really need to define our attitude profile every time we ask for the control input? Similar issue with checking for the control algorithm type at every call of `run`. IMO the objectives and algorithms of the controller objects should be given clear distinctions on startup, and inherit from a common set of interfacing methods.

Reply:
the function isnt used anymore, we can ditch it. I thought we'd need a guidance algorithm, but we can get away with having everything in the controller.

Reply:
function has been ditched